In [1]:
import pandas as pd
import numpy as np
import itertools
from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.linear_model import PassiveAggressiveClassifier
from sklearn.metrics import accuracy_score, confusion_matrix

In [3]:
# read the data

df = pd.read_csv('/home/lennox/notebooks/news.csv')

print(df.shape)

df.head()

(6335, 4)


Unnamed: 0                                              title   
0        8476                       You Can Smell Hillary’s Fear  \
1       10294  Watch The Exact Moment Paul Ryan Committed Pol...   
2        3608        Kerry to go to Paris in gesture of sympathy   
3       10142  Bernie supporters on Twitter erupt in anger ag...   
4         875   The Battle of New York: Why This Primary Matters   

                                                text label  
0  Daniel Greenfield, a Shillman Journalism Fello...  FAKE  
1  Google Pinterest Digg Linkedin Reddit Stumbleu...  FAKE  
2  U.S. Secretary of State John F. Kerry said Mon...  REAL  
3  — Kaydee King (@KaydeeKing) November 9, 2016 T...  FAKE  
4  It's primary day in New York and front-runners...  REAL

In [4]:
# get the labels 

labels = df.label
labels.head()

0    FAKE
1    FAKE
2    REAL
3    FAKE
4    REAL
Name: label, dtype: object

In [5]:
# split the data into train and test sets

x_train, x_test, y_train, y_test = train_test_split(df['text'], labels, test_size=0.2, random_state=7)

In [6]:
# lets initialize TfidfVectorizer,
# TfidfVectorizer takes a collection of text documents as input and transforms them into a numerical representation suitable for machine learning algorithms.
# It computes the Term Frequency-Inverse Document Frequency (TF-IDF) value for each term in the document collection.

tfidf_vectorizer = TfidfVectorizer(stop_words='english', max_df=0.7)

# Fit and transform train set, transform test set

tfidf_train = tfidf_vectorizer.fit_transform(x_train)
tfidf_test = tfidf_vectorizer.transform(x_test)


In [8]:
# initialize passiveAggressiveClassifier

pac = PassiveAggressiveClassifier(max_iter=50)
pac.fit(tfidf_train, y_train)

# predict using test set and calculate accuracy
y_pred = pac.predict(tfidf_test)
score = accuracy_score(y_test, y_pred)
print(f'Accuracy: {round(score*100, 2)}%')

Accuracy: 92.74%


we got an accuaracy of 92.74, quite impressive

In [9]:
# lets build a confusion matrix

confusion_matrix(y_test, y_pred, labels=['FAKE', 'REAL'])

array([[588,  50],
       [ 42, 587]])

We have 588 True positives, 567 True negatives, 50 false positives and 42 false negatives